Turn order

DanUgore opened the pull request:
This is just to show exactly what I added for turn order. Not to be merged.

DanUgore committed:
Add a queue sorting method.

It doesn't modify the actual queue, only simulates it for testing
purposes.

Commit:
Reset `tieOrder` each turn

DanUgore committed:
Add some comments

DanUgore committed:
Remove `runSwitch` action.

It did nothing but resort the queue unnecessarily. Also fix some debug
messages.

DanUgore committed:
Fix crashes and errors in `sortQueue()`

It would not place elements at the beginning and would crash if
non-pokemon related decisions tied.

DanUgore committed:
Use `sortQueue()` to sort actions, remove debugs

DanUgore committed:
Use `sortQueue()` instead of `addQueue(null)`

Also remove some irrelevant comments

DanUgore committed:
Readd 'runSwitch' action.

This was needed for something after all.

DanUgore committed:
Prevent sorting the queue prematurely

We don't sort the queue until all the active Pokemon are sent out. This
way we don't resolve speed ties too early.

DanUgore committed:
Fix errors in `sortQueue()`

I had forgotten to break the loop after the speed tie and to push
decisions decided by pre-existing tie order to the queue. Fixed here.

DanUgore commented:
@Zarel Here it is, seems fully functional now for actions. Haven't looked at sorting residuals with this yet.

DanUgore commented on battle-engine.js:
```diff
@@ -3173,6 +3177,7 @@ Battle = (function () {
 			if (!decision.pokemon && !decision.speed) decision.speed = 1;
 			if (!decision.speed && decision.choice === 'switch' && decision.target) decision.speed = decision.target.speed;
 			if (!decision.speed) decision.speed = decision.pokemon.speed;
+			if (decision.choice === 'move' || decision.choice === 'switch') decision.speed = decision.speed & 8191; // Silly GameFreak shenanigans.
```
Do you think I should hardcode Trick Room here btw?

Zarel commented:
I still don't know what's wrong with just using a stable sort. :|

Zarel commented:
Why do we even sort mid-turn, anyway? In theory, it should never happen.

DanUgore commented:
Sorry I still don't understand how to sort stably here. When we sort we need to resolve speed-ties, which we cant do if we sort before all pokemon are present.

Zarel commented:
So, why can't we sort after all pokemon are present?

DanUgore commented:
This does. In the 'start' event, `switchIn()` would sort after every call and it was called once for every Pokemon as it came in on Turn 0. That's no longer an issue,

As for sorting mid-turn, there is the move Quash to deal with. That's all I can think of.

Zarel commented:
Quash can manually change turn order, it doesn't need to trigger a full sort.

DanUgore commented:
Yea I just saw that

Zarel commented:
Basically, sort should happen at the beginning of every turn, and right after all Pokémon are switched in during turn 0. `BattlePokemon.prototype.tieOrder` should not exist. Anything else?

DanUgore commented:
Wait, why should that not exist?

DanUgore commented:
Well I guess it's actually possible/better to keep track of that in the battle object as opposed to each pokemon/

Zarel commented:
No, I mean, you should never have to keep track of it, since we should only ever sort once, at which point ties should be resolved instantly, right?

DanUgore commented:
Oh I see, well there's that and we need sort in the same order for end-of-turn residuals.

Zarel commented:
Oooooooh.

Yeah, keep `.tieOrder`.

Zarel commented:
Wait, how do switches work re: residuals?

DanUgore commented:
That is one of the things on my "to investigate" list.

Zarel commented:
You might want to quash this. At least the "fix" commits.

DanUgore commented:
K
- V4

> On Sep 7, 2014, at 1:45 AM, Guangcong Luo notifications@github.com wrote:
> 
> You might want to quash this. At least the "fix" commits.
> 
> \
> Reply to this email directly or view it on GitHub.

DanUgore commented:
Well apparently there is a lot of rerolling of speed ties.
1) Before the turn plays out
2) After Mega Evolution
3) Before residuals

So I wonder if we even need to keep track of tieOrder if we don't resort without rerolling after it's dealt with.

DanUgore commented:
Ok so it looks like we can get rid of tie order. Everywhere it's resorted it rerolls so we'll see how this turns out.

Zarel commented:
Okay, good, that makes things simpler.

DanUgore committed:
Reroll speed ties when needed

Add a 'resort' decision to the queue to reroll speedties right before moves (after mega evolution)
We sort the queue without rerolling after a switch, but the decisions are already in the proper order from the last sort.

Zarel commented:
So... is there anything that still needs to be done here?

DanUgore commented:
Thanks for reminding me. Mega evolution needs to be it's own queue thing. Speed is refilled post-mega evolution. Winning the mega speed tie doesn't guarantee you the same when you move.

> On Jun 12, 2015, at 1:55 PM, Guangcong Luo notifications@github.com wrote:
> 
> So... is there anything that still needs to be done here?
> 
> ―
> Reply to this email directly or view it on GitHub.

DanUgore commented:
Will probably look into this again. The fact Im actually approached a few times about double speed ties to this day is actually kind of dumb. Lost the local repo so Ill have to do some work over.
